### ISSUE TITLE
openssl s_server never sends a TLS close_notify

### ISSUE THREAD

**Issue by markdascher** [openssl/openssl#1806]
`openssl s_server` never sends a TLS **close_notify**.  This might be intentional, but it makes it impossible to test certain scenarios.  Specifically, a (possibly buggy) TLS server can respond with any of the following after receiving a **close_notify**:
- **close_notify** and TCP close
- **close_notify** without TCP close
- immediate TCP close
- nothing

The first two scenarios can't be tested with `openssl s_server`.  Instead, it closes the TCP connection immediately . [s_server.c](https://github.com/openssl/openssl/blob/master/apps/s_server.c) never calls [SSL_shutdown](https://www.openssl.org/docs/manmaster/man3/SSL_shutdown.html).

**Comment by mattcaswell** [comment]
I believe this is intentional (there is code to explicitly mark a connection as shutdown even though SSL_shutdown() has never been called). I'm not sure why this is the case and probably calling SSL_shutdown is better (or at least there should be an option to do it). However probably this is not needed for the 1.1.1 release, so moving to a post-1.1.1 milestone.

**Comment by t8m** [comment]
Please note we are looking for community help implementing this issue. Please comment if you are interested in implementing this feature or if you have reasons why this feature is important enough so we should consider implementing it with higher priority than other requested features. The issue will be closed after 3 months otherwise.

**Comment by markdascher** [comment]
Well, don't all speak up at once, eh? 😅

I admit it's hard to justify adding a feature without seeing widespread demand. Or in this case, any demand whatsoever… I seem to be the only one who would use it. And I don't even remember what I wanted to use it for, at this point.

I guess the only possible argument I can make is that, if even one person uses this feature to test SSL implementations, it can result in things like [this](https://github.com/rsyslog/rsyslog/issues/1231) or [that](https://github.com/netty/netty/issues/6167) being fixed, increasing the overall quality of implementations out there.

Basically I wonder if this is a chicken/egg problem. No one is aware of these quirks because there's no tool to test it. And no one is motivated to make a tool because they aren't aware of the quirks.

Also, I'm no SSL expert, so I don't really know what the significance of TLS close_notify is beyond its theoretical protection against truncation attacks. But I don't know what that really means in practice, or how common it is to actually implement properly.

But with all that said, I obviously wouldn't blame anyone for choosing not to dive into this obscure 10-year old edge case…

**Comment by igus68** [comment]
It looks like it was already done at some point. If we run TLS tests such as 70-test_renegotiation.t, 70-test_comp.t and others, they output messages sent between s_client and s_server, and we can see in the output many messages as

 Record 1 (server -> client)
  Content type: ALERT
  Version: TLS1.2
  Length: 52
  [1, 0]

According to RFC 5246 and RFC 8446, [1, 0] means [warning, close_notify], so s_server does send close_notify at least quite often.

**Comment by beldmit** [comment]
I believe it was fixed in https://github.com/openssl/openssl/pull/11734

**Comment by igus68** [comment]
As we have a merged PR and tests show that s_server now sends close_notify, I close this issue.
